File: cpp/leetcode_150_evaluate_reverse_polish_notation.cpp
class Solution {
public:
    int evalRPN(vector<string>& t) {
        std::vector<std::string>st;
        for(size_t i=0;i<t.size();){
            if(isdigit(t[i].front()) || t[i].size()>1){
                st.push_back(t[i++]);
            }else if(!isdigit(t[i].front()) && t[i].size()==1){
                int y=std::stoi(st.back());
                st.pop_back();
                int x=std::stoi(st.back());
                st.pop_back();
                switch(t[i].front()){
                    case '+': st.push_back(std::to_string(x+y));break;
                    case '*': st.push_back(std::to_string(x*y));break;
                    case '/': st.push_back(std::to_string(x/y));break;
                    case '-': st.push_back(std::to_string(x-y));break;
                }
                ++i;
            }
        }
        return std::stoi(st.back());
    }
};

///////////////////////
class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        stack<string> st; int opd1, opd2;
        int n = tokens.size();

        for (int i = 0; i < n; i ++) {
            if (tokens[i] == "+") {
                opd1 = stoi(st.top()); st.pop();
                opd2 = stoi(st.top()); st.pop();
                st.push(to_string(opd1 + opd2));
            } else if (tokens[i] == "-") {
                opd1 = stoi(st.top()); st.pop();
                opd2 = stoi(st.top()); st.pop();
                st.push(to_string(opd2 - opd1));
            } else if (tokens[i] == "*") {
                opd1 = stoi(st.top()); st.pop();
                opd2 = stoi(st.top()); st.pop();
                st.push(to_string(opd1 * opd2));
            } else if (tokens[i] == "/") {
                opd1 = stoi(st.top()); st.pop();
                opd2 = stoi(st.top()); st.pop();
                st.push(to_string(opd2 / opd1));
            } else {
                st.push(tokens[i]);
            }

        }

        return stoi(st.top());
    }
};

///////////////////

File: java/leetcode_150_evaluate_reverse_polish_notation.java
import java.util.regex.*;

class Solution {
    private Pattern pattern = Pattern.compile("-?\\d+(\\.\\d+)?");

    public boolean isNumeric(String strNum) {
        if (strNum == null) {
            return false; 
        }
        return pattern.matcher(strNum).matches();
    }
    public int evalRPN(String[] tks) {
        Stack<String>st=new Stack<>();
        for(String s:tks){
            if(st.size()==0 ||
             (isNumeric(st.peek()) && isNumeric(s)) )  st.push(s);
             else if(!isNumeric(s)){
                int y=Integer.parseInt(st.pop());
                int x=Integer.parseInt(st.pop());
                String res;
                switch(s){
                    case "*":   res=x*y+"";
                                st.push(res);
                    break;
                    case "+":   res=x+y+"";
                                st.push(res);
                    break;
                    case "/":   res=x/y+"";
                                st.push(res);
                    break;
                    case "-":   res=x-y+"";
                                st.push(res);
                    break;
                }
             }

        }
        return Integer.parseInt(st.peek());
    }
}

//////////////////
class Solution {
    public int evalRPN(String[] tokens) {
        Stack<Integer> st = new Stack<Integer>();
        
        for (String token : tokens) {
            if (token.equals("+")) {
                int b = st.pop();
                int a = st.pop();
                st.push(a + b);
            } else if(token.equals("-")){
                 int b = st.pop();
                int a = st.pop();
                st.push(a - b);
            }else if(token.equals("*")){
                int b = st.pop();
                int a = st.pop();
                st.push(a * b);
            }else if(token.equals("/")){
                 int b = st.pop();
                int a = st.pop();
                st.push(a / b);
            }else{
                st.push(Integer.parseInt(token));
            }
        }
        int i = st.pop();
        return i;
    }
}

/////////////////
